core/src/mast/mod.rs

```diff
@@ -60,6 +62,82 @@ impl MastForest {
         Ok(new_node_id)
     }
 
+    /// Adds a basic block node to the forest, and returns the [`MastNodeId`] associated with it.
+    pub fn add_block(
+        &mut self,
+        operations: Vec<Operation>,
+        decorators: Option<DecoratorList>,
+    ) -> Result<MastNodeId, MastForestError> {
+        match decorators {
+            Some(decorators) => {
+                self.add_node(MastNode::new_basic_block_with_decorators(operations, decorators))
+            }
+            None => self.add_node(MastNode::new_basic_block(operations)),
+        }
+    }
+
+    /// Adds a join node to the forest, and returns the [`MastNodeId`] associated with it.
+    pub fn add_join(
+        &mut self,
+        left_child: MastNodeId,
+        right_child: MastNodeId,
+    ) -> Result<MastNodeId, MastForestError> {
+        match self.add_node(MastNode::new_join(left_child, right_child, self))? {
+            node_id if node_id <= left_child || node_id <= right_child => {
+                Err(MastForestError::InvalidNodeId(node_id))
+            }
+            node_id => Ok(node_id),
+        }
```

Comment:
I think the invalid node ID here is not the `node_id` but either `left_child` or `right_child` - no?